src/array.jl

```diff
@@ -552,18 +552,66 @@ function copyto!(dest::CatArrOrSub{T, N, R}, dstart::Integer,
     dest
 end
 
-copyto!(dest::CatArrOrSub, src::CatArrOrSub) =
+function copyto!(dest::CatArrOrSub{T1, N, R}, dstart::Integer,
+                 src::AbstractArray{T2,N}, sstart::Integer,
+                 n::Integer) where {T1, T2, N, R}
+    n == 0 && return dest
+    n < 0 && throw(ArgumentError(string("tried to copy n=", n, " elements, but n should be nonnegative")))
+    destinds, srcinds = LinearIndices(dest), LinearIndices(src)
+    (checkbounds(Bool, destinds, dstart) && checkbounds(Bool, destinds, dstart+n-1)) || throw(BoundsError(dest, dstart:dstart+n-1))
+    (checkbounds(Bool, srcinds, sstart)  && checkbounds(Bool, srcinds, sstart+n-1))  || throw(BoundsError(src,  sstart:sstart+n-1))
+    srclevs = DataAPI.refpool(src)
+    srcrefs = DataAPI.refarray(src)
+    # Fast path only supports refs which can be used to index
+    # into vector mapping from source levels to destination levels
+    if !(srclevs isa AbstractVector) || !(srcrefs isa AbstractArray{<:Integer})
+        return invoke(copyto!, Tuple{AbstractArray, Integer, AbstractArray, Integer, Integer},
+                      dest, dstart, src, sstart, n)
+    end
+    checkbounds(dest, axes(src)...)
+    if !(T1 >: Missing) && any(ismissing, srclevs)
```

Comment:
`srclevs` might contain `missing` but the copied region might not contain them. Why do we throw error in this case?

Reply:
That's a very good point. Actually that's a general problem with pools that can include unused values (https://github.com/JuliaData/PooledArrays.jl/issues/36). I had to refactor things quite a bit to take this into account. This should make things slower when there are unused levels that are not at the end, but luckily CSV.jl doesn't generate this kind of object.